### DISCUSSION TITLE
SSE: Why we close `eventSink` twice?

### DISCUSSION THREAD

**Discussion by kkazuo** [resteasy/resteasy#3763] [General]
Why we close `eventSink` twice? (L66, L68)

https://github.com/resteasy/resteasy/blob/082852c84d5c546d3ff44bb4f82d0da5325e115b/resteasy-core/src/main/java/org/jboss/resteasy/plugins/providers/sse/SseBroadcasterImpl.java#L65-L74

Thank you.

**Comment by liweinan** [comment]
I remember this is to ensure the `eventSink` is closed to avoid some racing conditions. The implementation of the `EventSink` is here:

- https://github.com/resteasy/resteasy/blob/main/resteasy-core/src/main/java/org/jboss/resteasy/plugins/providers/sse/SseEventOutputImpl.java

And the `close` method contains some async operations:

- https://github.com/resteasy/resteasy/blob/082852c84d5c546d3ff44bb4f82d0da5325e115b/resteasy-core/src/main/java/org/jboss/resteasy/plugins/providers/sse/SseEventOutputImpl.java#L385C15-L385C15

**Reply by jamezp** [reply]
IMO this looks like a bug. If the first `close()` fails the notifications would never happen. We should only be closing this once and if there is a race condition, we need to fix that in a better way.

**Comment by liweinan** [comment]
> IMO this looks like a bug. If the first `close()` fails the notifications would never happen. We should only be closing this once and if there is a race condition, we need to fix that in a better way.

+1 I remember this was added to solve random test failures.

**Comment by jamezp** [comment]
I've filed https://issues.redhat.com/browse/RESTEASY-3369 for this.
